### ISSUE TITLE
Custom WebFinger domain (à la Mastodon's `LOCAL_DOMAIN`)

### ISSUE THREAD

**Issue by tesaguri** [kitsune-soc/kitsune#285]
Mastodon uses (misleadingly named IMO) [`LOCAL_DOMAIN`] environment variable value to set the domain part of canonical WebFinger `subject`s of the server. It would be great if Kitsune supported a similar configuration.

[`LOCAL_DOMAIN`]: <https://docs.joinmastodon.org/admin/config/#local_domain>

Suppose Alice is running a  Mastodon server at `social.alice.example` domain. Naively, Alice's handle name would be `@alice@social.alice.example`, but why bother with the `social.` subdomain? In the world of e-mail, Alice can run an SMTP server at `smtp.alice.example` and receive mails as `alice@alice.example` by setting appropriate DNS records. `LOCAL_DOMAIN` works just like that: Alice's account will be recognized as `@alice@alice.example` by setting `LOCAL_DOMAIN` to `alice.example` (and setting up appropriate HTTP redirections in `alice.example` domain), while the actual Mastodon server is running at `social.alice.example`.

[`social.treehouse.systems`](<https://social.treehouse.systems>) is an example of such a Mastodon server. When you look up <https://social.treehouse.systems/@ariadne> in another Mastodon server, the account will be shown as `@ariadne@treehouse.systems` instead of `@ariadne@social.treehouse.systems`. But how does the other Mastodon server know that `@ariadne@treehouse.systems` is the canonical handle?

Let's follow the discovery process by hand. First, fetch the Activity Streams object at <https://social.treehouse.systems/@ariadne>, which tells us that `ariadne` is the actor's `preferredUsername`. Then, perform a WebFinger query <https://social.treehouse.systems/.well-known/webfinger?resource=acct:ariadne@social.treehouse.systems>, and we get a JRD document with its `subject` field being `acct:ariadne@treehouse.systems`. But since the domain part of the `subject` is different from the ActivityPub actor's domain, we need to verify with `treehouse.systems` by performing another WebFinger query <https://treehouse.systems/.well-known/webfinger?resource=acct:ariadne@treehouse.systems>, just in case `social.treehouse.systems` is making a bogus claim to the authority over the other domain. If the second WebFinger query result agrees with the first one (or simply redirects back to <https://social.treehouse.systems/.well-known/webfinger?resource=acct:ariadne@treehouse.systems>), then we can finally confirm that `acct:ariadne@treehouse.systems` is the canonical subject!

To sum up, in order to support a configuration like `LOCAL_DOMAIN`, the WebFinger handler needs to accept both the server's actual domain and the configured domain. I haven't read the code throughly, but some code depending on `kitsune::service::UrlService::domain` might also need to be updated. At least, [`uri` field of `V1::Instance`][v1-instance-uri] should use the `LOCAL_DOMAIN` value.

[v1-instance-uri]: <https://docs.joinmastodon.org/entities/V1_Instance/#uri>

I think this might also be a good fit for Shuttle's Hobby plan where you cannot bring your own domain (cf. #78). Although you would still be locked into Shuttle's domain in terms of ActivityPub object IDs, at least the handle names would look much better this way.

**Comment by aumetra** [comment]
That is definitely an interesting suggestion.

How would that interact with something like #179?

Having a webfinger resource on "example.com" and having it point to users on "social.example.com" should have already been possible, since our webfinger resolver just follows the link inside the resource and doesn't to any same-domain checks.

It sounds like this issue here has some overlap with #179 on a quick glance

**Comment by tesaguri** [comment]
> How would that interact with something like [#179](https://github.com/kitsune-soc/kitsune/issues/179)?

I think each virtual tenant (or "virtual instance") would (optionally) have its own `LOCAL_DOMAIN`-like value. The configuration might look like the following:

```toml
[[instance]]
name = "Alice's Kitsune instance"
# The configuration name is tentative
local_domain = "alice.example"

[[instance.url]]
scheme = "http"
domain = "social.alice.example"

[[instance]]
name = "Bob's Kitsune instance"
local_domain = "bob.example"

[[instance.url]]
scheme = "http"
domain = "social.bob.example"
```

Maybe the virtual instances would live in the database rather than the configuration, but the schema would look similar in that case too.

Under the above configuration, `https://social.alice.example/.well-known/webfinger?resource=acct:alice@social.alice.example` and `https://social.bob.example/.well-known/webfinger?resource=acct:bob@social.bob.example` should resolve to `acct:alice@alice.example` and  `acct:bob@bob.example` respectively.

The tricky part would be what the WebFinger handler should return when it received a query for an actor on a different virtual instance than the `HOST` header suggests, like <https://social.alice.example/.well-known/webfinger?resource=acct:bob@bob.example>. I think returning `acct:bob@bob.example` in that case would be just fine because I don't come up with a problem with that and the implementation would be significantly simpler that way.

Takahē seems to handle the case in the same way. For example, <https://jointakahe.takahe.social/.well-known/webfinger?resource=acct:admin@takahe.social> just returns `acct:admin@takahe.social`, which is on another virtual instance. There might be privacy concern if the administrator of the virtual instances doesn't want it known that they are operated in the same infrastructure, but that is already guessable by the IP address anyway.

> Having a webfinger resource on "example.com" and having it point to users on "social.example.com" should have already been possible, since our webfinger resolver just follows the link inside the resource and doesn't to any same-domain checks.

Yes, the current WebFinger resolver resolves such an actor just fine. But the current implementation doesn't seem to fetch and store the canonical WebFinger resource domain (`example.com` as opposed to `social.example.com`). This matters to the Mastodon API that should set the [`acct` field of the `Account` entity][mastodon-doc-entities-account-acct] for an actor at `social.example.com` to `{username}@example.com`, while the current implementation seems to return `{username}@social.example.com` (I haven't tested with a real Kitsune instance yet. Sorry if I'm misinterpreting the code). That's what the second todo in the OP is about (well, this might be complicated enough to be worth a separate issue?).

[mastodon-doc-entities-account-acct]: <https://docs.joinmastodon.org/entities/Account/#acct>

**Comment by aumetra** [comment]
> Maybe the virtual instances would end up living in the database rather than the configuration, but the schema would look similar in that case too.

Yep, seems reasonable. My idea was that users can bring their own domains and just register them on arbitrary instances if they choose to.  
Similar to how users can bring their own domains to something like Bluesky.

That's probably gonna be the next bigger thing I'll be looking into since that will be a big step into the direction of finally having a stable database schema that doesn't break each release.
